finality-aleph/src/party.rs

```diff
@@ -98,41 +139,109 @@ where
+    let mut last_block_seen = false;
+    let current_stop_h = session.stop_h;
+    loop {
+        tokio::select! {
+            maybe_batch = ordered_batch_rx.next() => {
+                if let Some(batch) = maybe_batch {
+                    trace!(target: "afa", "Received batch {:?} in aggregator.", batch);
+                    if last_block_seen {
+                        //This is only for optimization purposes.
+                        continue;
+                    }
+                    for new_hash in batch {
+                        let to_finalize_headers = chain_extension_step(last_finalized, new_hash, client.as_ref());
+                        for header in to_finalize_headers.iter() {
+                            if *header.number() <= current_stop_h {
+                                aggregator.start_aggregation(header.hash()).await;
+                                last_finalized = header.hash();
+                            }
+                            if *header.number() >= current_stop_h {
+                                aggregator.notify_last_hash().await;
+                                last_block_seen = true;
+                                break;
+                            }
+                        }
+                    }
+                } else {
+                    debug!(target: "afa", "Batches ended in aggregator.");
+                }
+            }
+            multisigned_hash = aggregator.next_multisigned_hash() => {
+                if let Some((hash, multisignature)) = multisigned_hash {
+                    let number = client.number(hash).unwrap().unwrap();
+                    // The unwrap might actually fail if data availability is not implemented correctly.
+                    let notification = JustificationNotification {
+                        justification: AlephJustification::new::<B>(multisignature),
+                        hash,
+                        number
+                    };
+                    if let Err(e) = justification_tx.unbounded_send(notification)  {
+                        error!(target: "afa", "Issue with sending justification from Aggregator to JustificationHandler {:?}.", e);
+                    }
+                    //finalize_block_as_authority(client.clone(), hash, multisignature);
+                } else {
+                    debug!(target: "afa", "The stream of multisigned hashes has ended. Terminating.");
+                    return;
+                }
+            }
+            _ = &mut exit_rx => {
+                debug!(target: "afa", "Aggregator received exit signal. Terminating.");
+                return;
+            }
+        }
+    }
 }
 
 /// If we are on the authority list for the given session, runs an
 /// AlephBFT task and returns `true` upon completion. Otherwise, immediately returns `false`.
-async fn maybe_run_session_as_authority<B, C, BE, SC>(
-    authority: AuthorityId,
+#[allow(clippy::too_many_arguments)]
```

Comment:
:<

I don't see how to fix this easily, but this almost surely needs at least a refactor in the future. With the zero bug policy I'm not convinced this can be ensured, so could you try fixing this somehow?

Reply:
The typical solution is to intruduce a `RunSessionParams` struct. This is technically equivalent, but is usually clearer because fields of a struct have names unlike function parameters

Reply:
Yeah, I'm aware of this fix, but I don't see how this makes the code better (we can get rid of `#[allow(clippy::too_many_arguments)]` but let's not count that :P) -- there would be then a few new "Params"-type structs defined for the sole purpose of silencing clippy.  Reading the code and (to a lesser extent) writing is then harder, not simpler.

Reply:
There is now a task for this in the backlog.